### DISCUSSION TITLE
AB Scripting - QC form if Request Type is Production

### DISCUSSION THREAD

**Discussion by hagen31** [sadieblue/agilityblue.workspace-scripts#16] [Q&A] [answered]
Hello,

I have a script to create a QC Task when a Production form is completed. I am looking to take that a step further where on the Production form if a field called Request Type (a single choice field) is Production I would like that to be the trigger for the new form. I've been researching using Label and Value; however, have not had any luck. Any guidance would be appreciated.

**Comment by bmward80** [comment]
HI @hagen31,

You want the Request Type to equal "Production" to create the new QC Task, is that right? Here's a rough order to accomplish this:

1. Ensure your script is setup to use a before save event trigger on the task object. It needs to be this type of trigger so you can compare existing instances vs incoming changes.
2. Grab a copy of the current state of the task using `Get-Task`. You would need to iterate through the task's forms, sections, and fields to locate the request type field (there's a question recently that answered how to do that [here](https://github.com/sadieblue/agilityblue.workspace-scripts/discussions/14)). You would then check that field's `ValueAsString` property to see if the "Production" value already exists or not. 
3. Compare the Request Type field between the existing one and the inbound one. The point here is that you only want to create a QC task if the inbound one is equal to "Production" and the existing one does not have that value, otherwise you might create qc tasks inadvertently each time the task was updated. You could even add an additional check perhaps closer to the start of the script that checks if the project already contains a qc task and bail out if it does as an added safety check.
4. If the previous step was satisfied, you'd move onto creating the qc task.

We don't have any examples on hand that are quite this specific, but this structure is the way I'd approach it. Let me know if you run into any roadblocks and I'll see if I can be of any further help.

Regards,
Brandon

**Reply by hagen31** [reply]
Hello Brandon,

I've been using that as a guide with no luck getting the field properties. I'll keep trying though.

**Reply by bmward80** [reply]
Do you have any code to share that you're stuck on specifically, or any errors you're experiencing so I can help?

**Reply by hagen31** [reply]
The snippet of code related to this question is below. Right now, I am just trying to get the script to return the value of the field. Once I get that figured out I will go into the broader code and can paste it all. I do not get an error rather the ouput is just a black box.

$task.Forms | ForEach-Object {
  $form = $_

  $form.Sections | ForEach-Object {
    $section = $_
 
    $myField = $section.Fields | Where-Object { $_.Label -eq "Request Type" }

      if ($field.Label -eq "Request Type") {
        Write-Output "The value is $($myField.ValueAsString)."
      }
    }
  }

**Reply by bmward80** [reply]
Thanks for this. There doesn't appear to be anything wrong with the code. Do you know how your `$task` variable is being populated? Either 2 things are happening: 1) The $task variable is not populated, or 2) there is no "Request Type" field in the task. You can convert an object to JSON to see all of the fields as a helpful way of debugging through a script. Something like the following should output your task variable:

```powershell
# Explicitly retrieve a task
$task = Get-Task -Id 1 -IncludeForms

# Write the output of the task as JSON structure to the console. You can then observe all of the available forms,
# sections, fields, etc.
$task | ConvertTo-Json -Depth 10
```

Brandon

**Reply by hagen31** [reply]
Thank you for this. I've tried with most of the fields on the task and using the above verified the $task variable is not populating as it returns null.

**Reply by bmward80** [reply]
Do you know how the `$task` variable should be populating? Are you retrieving a specific task by hard coding one, or passing a parameter in when running the script manually. Or are you triggering the script on an event? Do you have the code that assigns something to the `$task` variable?

**Reply by hagen31** [reply]
For the testing I am triggering off of and event and for now I want to output what the value of the Request Type is from the closed task. Your note above made me realize I was not giving the $task the correct variable value. I made an update to that and was able to get the expect results. I then put into my larger script and results are successful. I have another tweak I need to make to the larger script then I can post the full script for my ultimate goal which is as below.

Create a script to automatically create a Production QC Task when a Production Task is Request Type Production and email a group. If a Production QC Task is already created on the Project or Request Type is not Production, then do nothing. Optionally automatically assign the Production QC task to a specific group. I haven't decided on the last part so will likely comment out the code.

**Reply by hagen31** [reply, marked as answer]
``
``# ------------------------------------------------------------------------------
#
# When a user completes a task, this script will create a new task on the same
# project that uses a QC form.
#
# ------------------------------------------------------------------------------
# In addition to automating project and task creation based on forms,
# another benefit to using a script to create tasks is that the script can ignore
# required fields within the forms while creating the task. This allows you to
# create forms with required fields that are not required when the task is
# created from the form, but required when a user wants to update the task
# within the application.
#
# ------------------------------------------------------------------------------
# Event triggers to setup:
#   1) Object: Task, Action: On Update, Action State: After Save
#
# ------------------------------------------------------------------------------
# USER PARAMETERS

# The task form id is checked when a task is completed. If the task contains a form
# with this id, then a new task will be created using the QC form.
$taskFormId = 39 # Production Form

# The QC form id is used to create a new task when a task is completed, provided the
# task that was completed contains the taskFormId from above.
$qcFormId = 79 # Production QC Form

# The User ID to assign the task to. Complete as needed
$assignToUserId = ""
# The User name for this user. Complete as needed
$assignToUserName = ""

# Define the parameters for the notification
$requestType = "New Task"


# ------------------------------------------------------------------------------
# SCRIPT BODY

$taskId = Get-InboundObjectId

if ($null -eq $taskId) {
  # If the inbound object id is not available, it means we are executing the script manually.
  # In this case, we'll set an id directly for testing purposes
  $taskId = 10000
}

# Here, we are retrieving the task and including the forms so we can access 
# form properties.
$task = Get-Task $taskId -IncludeForms

# We can uncomment this line to output the properties/values available to us
# $task | ConvertTo-JSON

# We want to make sure that the task is closed
if ($task.StatusDescription -ne "Completed") {
  # The task is already closed out, so do nothing
  Write-Output "Task $($task.TaskId) is not completed. No action will be performed."
  Exit
}

# Check if the task contains the task form id we are looking for
if (-not ($task.Forms | Where-Object { $_.FormId -eq $taskFormId })) {
  # The task doesn't contain the form we're looking for, so do nothing
  Write-Output "Task $($task.TaskId) doesn't contain form $($taskFormId). No action will be performed."
  Exit
}

# Verify this is a Production export
$task.Forms | ForEach-Object {
    $form = $_

    $form.Sections | ForEach-Object {
        $section = $_

        # Output the labels of fields in this section for debugging
        $section.Fields | ForEach-Object {
        }

        # Find the field with the label "Request Type"
        $myField = $section.Fields | Where-Object { $_.Label -eq "Request Type" }

        # Check if the field was found and then output its value
        if ($myField.ValueAsString -eq "Production") {
            return
        } else {
            Write-Output "This is not a Production request, no QC form will be created."
            Exit
        }
    }
}

# Before we create a QC task, let's make sure that one doesn't already 
# exists in the project.

$qcTaskAlreadyExists = $false

$tasksInProject = Get-Tasks -Filter "ProjectId eq $($task.ProjectId)" -Top 1000

$tasksInProject.Collection | ForEach-Object {
  $taskInProject = Get-Task $_.TaskId -IncludeForms

  if ($taskInProject.Forms | Where-Object { $_.FormId -eq $qcFormId }) {
    $qcTaskAlreadyExists = $true
    return
  }
}

if ($qcTaskAlreadyExists) {
  # A QC task already exists in the project, so do nothing
  Write-Output "A QC task already exists in project $($task.ProjectId). No action will be performed."
  Exit
}

# Get the form that we want to use for the QC task
# Use the Agility Blue `Get-Form` PowerShell command followed by the 
# form id to retrieve the form
$qcForm = Get-Form $qcFormId

# Create a new task with the basic task information
$qcTask = @{
  Name      = $qcForm.Name
  ProjectId = $task.ProjectId
  DateDue   = $task.DateDue
  Forms     = @()
}

# To use a form for a task, we need to create a task form object that's
# based on the form we want to use. We'll loop through each section and
# field in the form and create a task form object that we can use to
# create the task.

# Forms contain sections, and sections contain fields.

$qcTaskForm = @{
  FormId   = $qcForm.FormId
  Sections = @()
}

$qcForm.Sections | ForEach-Object {
  $qcFormSection = $_

  $qcTaskFormSection = @{
    FormSectionId = $qcFormSection.SectionId
    Fields        = @()
  }

  $qcTaskForm.Sections += $qcTaskFormSection

  $qcFormSection.Fields | ForEach-Object {
    $qcFormSectionField = $_

    $qcTaskFormSectionField = @{
      FieldId = $qcFormSectionField.FieldId
    }

    # Check if there is a default value on the field. If there is, apply the default
    # To the appropriate ValueAs* field
    if (-not [string]::IsNullOrWhiteSpace($qcFormSectionField.DefaultValue)) {
      switch ($qcFormSectionField.DataTypeName) {
        "Whole Number" {
          # Convert the DefaultValue to an integer
          $qcTaskFormSectionField['ValueAsNumber'] = `
            [int]$qcFormSectionField.DefaultValue
        }
        "Decimal Number" {
          # Convert the DefaultValue to a decimal
          $qcTaskFormSectionField['ValueAsDecimal'] = `
            [decimal]$qcFormSectionField.DefaultValue
        }
        "Yes or No Choice" {
          # Convert the DefaultValue to a boolean
          $qcTaskFormSectionField['ValueAsBoolean'] = `
            [System.Convert]::ToBoolean($qcFormSectionField.DefaultValue)
        }
        { ($_ -eq "Date Only") -or ($_ -eq "Date and Time") } {
          # Convert the DefaultValue to a date
          $qcTaskFormSectionField['ValueAsDate'] = `
            [DateTimeOffset]$qcFormSectionField.DefaultValue
        }
        { ($_ -eq "Single Choice") -or ($_ -eq "Multiple Choice") } {
          # These fields don't have defaults, so they can be ignored unless you want
          # to add selected options during the task creation.
          break;
        }
        default {
          $qcTaskFormSectionField['ValueAsString'] = `
            $qcFormSectionField.DefaultValue
        }
      }
    }

    # Reference fields need to be handled in a different way. They exist as a
    # separate "ReferenceObject" property on the field and contain a list of
    # Values.
    if (($qcFormSectionField.DataTypeName -eq "Reference") `
        -and ($null -ne $qcFormSectionField.ReferenceObject -ne $null)) {

      # Create the reference object property that requires the object id and
      # the list of values.
      $qcTaskFormSectionField['ReferenceObject'] = @{
        ObjectId = $qcFormSectionField.ReferenceObject.ObjectId
        Values   = @()
      }

      # If the form has default reference values, we'll add them here
      $qcFormSectionField.ReferenceObject.DefaultValues | ForEach-Object {
        $refVal = @{
          KeyAsInteger = $_.KeyAsInteger
          KeyAsLong    = $_.KeyAsLong
        }

        # Again, PowerShell will convert $null string properties to an empty string,
        # so we'll handle objects that use a string as an ID separately here.
        if (-not [string]::IsNullOrWhiteSpace($_.KeyAsString)) {
          $refVal["KeyAsString"] = $_.KeyAsString
        }

        # Add the ference values to the reference object
        $qcTaskFormSectionField.ReferenceObject.Values += $refVal
      }
    }

    # Add the field to the section
    $qcTaskFormSection.Fields += $qcTaskFormSectionField
  }
}

# Finally, add the form to the task
$qcTask.Forms += $qcTaskForm

# Uncomment this line to view the task object that will be created
# $qcTask.Forms | ConvertTo-Json -Depth 10

# Use the Agility Blue `Add-Task` PowerShell command followed by the 
# task object to create the task
$createdTask = Add-Task $qcTask

# Use the Agility Blue `Assign-Task` PowerShell command followed by the 
# task object to assign the task
$task = Set-AssignTask -TaskId $createdTask.TaskId -UserId $assignToUserId

# Create a notification using the Invoke-Notification function
Invoke-Notification -RequestType $requestType -Id $createdTask.TaskId

Write-Output "Created QC task $($createdTask.TaskId) in project $($createdTask.ProjectId) and assigned to $($assignToUserName)."
###

**Reply by hagen31** [reply]
The code didn't paste very well; however, it is above.

**Reply by bmward80** [reply]
Very cool, @hagen31, glad that you have made good progress! 

If you're interested in formatting code better here, Github uses "markdown" formatting. If you want to wrap your powershell code in a comment to make it look "pretty", you can surround your code between 3 tick mark characters `, like this:

![image](https://github.com/user-attachments/assets/7ca42990-a40d-43da-ab11-0bf9596fd806)

It ends up looking like this:

```powershell
$myVar = "Your powershell code here"
```

You can see a "Markdown is supported" link below on the bottom right of the comment and you can also use the preview tab to see how it would look before replying.

Have a good weekend!
Brandon

**Reply by hagen31** [reply]
I'll look into that. What I am learning about Powershell is there is a lot of declaring the variables so not worrying about being pretty at the moment. Working through getting use to that.

Have a good weekend as well.

**Reply by bmward80** [reply]
Sure, I understand that. What I meant was for sharing code here on this forum, not anything directly with PowerShell.
